### DISCUSSION TITLE
Proposal for NON_START_CHAR

### DISCUSSION THREAD

**Discussion by hchiba1** [pg-format/pg-formatter#11] [Ideas]
Proposed by @nichtich :

Plain values must not start with " (otherwise they must be quoted strings) and with opening parenthesis (to not confuse with syntax of YARS-PG and Cypher and to possibly allow for extension of PG syntax).

Single quoted strings are also removed because these are less likely required and to be equal to JSON strings. Unicode escape sequences are not included yet.

```diff
$ diff lib/pg.2024-02-24.pegjs lib/pg.2024-02-25.pegjs
< SingleQuotedChar = !("'" / "\\") char:.
< {
<   return char;
< }
< / "\\" esc:EscapedChar
< {
<   return esc;
< }
< 
150,157c141
< / "'" chars:SingleQuotedChar* "'"
< {
<   return {
<     quote: "'",
<     literal: chars.join(''),
<   };
< }
< / chars:NON_SPECIAL_CHAR+
---
> / first:NON_START_CHAR chars:NON_SPECIAL_CHAR*
161c145
<     literal: chars.join(''),
---
>     literal: first + chars.join(''),
173a158
> NON_START_CHAR = [^:"(\x20\x09\x0D\x0A]
```

```diff
$ diff -w test/parser.js.2024-02-24 test/parser.js.2024-02-25
23a24,25
> // TODO:
> // '"\\u1234"' Unicode escape sequence
45d46
< /* These need to be detected:
49c50
<   'a :\'\\"',  // missing end of quoted string with escaped '
---
>   'a :\"\\"',   // missing end of quoted string with escaped '
52,54c53,54
< /* These need to be discussed:
<   '(a',       // invalid id (must not start with bracket)        
<   'a)',       // invalid id (must not end with bracket) */
---
>   '(a',         // invalid id (must not start with opening parenthesis)       
>   '"\\uxxxx"',  // invalid Unicode Escape sequence
```

**Comment by nichtich** [comment]
The issue of non-start-char is which characters to allow in unquoted names (node identifiers, labels, property keys, and property values).

- space and tabulator is not allowed because it is whitespace
- `"` is not allowed because it starts a quoted name
- `#` is not allowed because it starts a comment
    - Edge cases label `:#` and property value `key:#value` may be allowed (what do you think?)
- `:` should not be allowed because it starts a label.
    - Edge cases label `::x` (label is `":x"`) may be allowed, what do you think?
- `(` should not be allowed to support extension of PG format with queries, templates - extensions are out of the scope of PG format but there should be some room to define a proper superset of the format based on parenthesis expressions.

Note that some of these characters may be allowed inside names (and all characters are allowed in quoted names anyway).

**Comment by hchiba1** [comment]
I introduced `UNQUOTED_CHAR` (05dd5ed).

`:#` and `key:#value` are still allowed.
(Note that trailing comments currently require space before `#`)

While I met minimal requirements, `UNQUOTED_CHAR` can be further extended.

**Comment by nichtich** [comment]
Having read the CYPHER specification I'd change START_CHAR to Unicode character classes `ID_Start` and `Pc` (basically letters and underscore) but then allow every character except for space, tabulator, " < > { } | ^ as following characters.

**Comment by hchiba1** [comment]
Followed by a general discussion:
* #37
